### ISSUE TITLE
torch.func transforms silently return zeros under inference_mode

### ISSUE THREAD

**Issue by stmcgovern** [pytorch/pytorch#177318]
## 🐛 Describe the bug

  `torch.func.{grad, vjp, jvp}` silently produce incorrect results (zeros)
  when called under `torch.inference_mode()`. No error is raised.

  ```python
  import torch
  from torch.func import jvp

  x = torch.randn(3)
  with torch.inference_mode():
      out, tangent = jvp(lambda x: x**2, (x,), (torch.ones(3),))
      print(tangent)  # tensor([0., 0., 0.]) — should be 2*x
```
##  Root cause

  Functorch transforms set up their own autograd context on entry:
  enable_grad() for grad/vjp, _set_fwd_grad_enabled(True) for jvp
  (see NOTE [grad and vjp interaction with no_grad]). However,
  inference_mode also excludes autograd dispatch keys from the TLS
  dispatch key set, and the mode-enabling calls don't undo that. So
  autograd dispatch keys remain excluded and differentiation silently
  produces zeros.

  The transforms already correctly handle no_grad — this is the same
  class of issue with a different TLS mechanism.

##  Expected behavior

  Either produce correct results (by also disabling inference_mode on
  transform entry, matching the existing no_grad handling) or raise
  an error.

cc @ezyang @gchanan @kadeng @msaroufim @albanD @gqchen @nikitaved @soulitzer @Varal7 @bobrenjc93 @Chillee @samdow @kshitij12345

**Comment by Lillian68** [comment]
Hi @stmcgovern and @soulitzer ,

I totally understand that it would be convenient for use cases like molecular dynamics, sensitivity analysis, and ODE solvers if these function transforms could work under torch.inference_mode().

However, I believe we should raise an explicit error instead of silently returning incorrect zeros or trying to make them work.

torch.inference_mode() is explicitly designed for inference-only workloads with autograd fully disabled. Running gradient-related transforms like grad, vjp, and jvp inside this mode is semantically inconsistent. Allowing them to run silently not only produces wrong results but also violates the design contract of inference_mode.

For users who need differentiation while avoiding unnecessary gradient tracking, torch.no_grad() is already the appropriate choice.

Therefore, I think the correct fix is to detect when these APIs are used under inference_mode() and raise a clear error message to prevent silent failures.
